community/rfcs/techniques/proposals/20200205-weight-clustering-proposal.md

```diff
@@ -0,0 +1,82 @@
+
+## Proposed Design
+
+The weight clustering technique that forms the basis of our implementation has been described in the conference paper *Deep Compression: Compressing Deep Neural Networks with Pruning, Trained Quantization and Huffman Coding* by S. Han, H. Mao, and  W. J. Dally, which can be downloaded from [arXiv](https://arxiv.org/abs/1510.00149). Please refer to chapter 3, entitled *Trained Quantization and Weight Sharing*, for details about the clustering technique &mdash; since the paper is brief and freely available for download, we will not reproduce its content here.
+
+From a practical standpoint, the implementation uses a lookup table to hold the cluster centroid values during model training. The weight array is populated with a 'gather' operation so that during back propagation the gradients can be calculated in the normal way. The lookup table is then adjusted using the cumulative gradient values for the weights that correspond to the same centroid.
+
+The number of unique values required, as well as the way the cluster centroids are initialized are passed in as parameters.
+
+The initial values of cluster centroids are fine-tuned during a subsequent training step.
+
+## Experimental Results
+
+The table below summarizes the compression rate and top-1 accuracy achieved when clustering the weights of different models using linear centroid initialization:
+
+<table>
+  <tr>
+    <th rowspan=2>Model</th>
+    <th rowspan=2>Dataset</th>
+    <th colspan=2>Original</th>
+    <th colspan=3>Clustered</th>
+  </tr>
+  <tr>
+    <th>Top-1 accuracy</th>
+    <th>Size of compressed .tflite</th>
+    <th># of clusters</th>
+    <th>Top-1 accuracy</th>
+    <th>Size of compressed .tflite</th>
+  </tr>
+  <tr>
+    <td>ConvNet</td>
+    <td>MNIST</td>
+    <td>99.40%</td>
+    <td>0.6 MB</td>
+    <td>32</td>
+    <td>98.78%</td>
+    <td>0.1 MB</td>
+  </tr>
+  <tr>
+    <td>MobileNetV2</td>
+    <td>ImageNet</td>
+    <td>72.29%</td>
+    <td>13.0 MB</td>
+    <td>32</td>
+    <td>69.33%</td>
+    <td>2.6 MB</td>
+  </tr>
+  <tr>
+    <td>DS-CNN-L</td>
+    <td>Speech Commands v0.02</td>
+    <td>95.03%</td>
+    <td>1.5 MB</td>
+    <td>32</td>
+    <td>94.71%</td>
+    <td>0.3 MB</td>
+  </tr>
+</table>
+
+NOTE: *Size of compressed .tflite* refers to the size of the zipped .tflite file obtained from the model through the following process:
+1. Serialize the Keras model into .h5 file
+2. Convert the .h5 file into .tflite using `TFLiteConverter.from_keras_model_file()`
+3. Compress the .tflite file into a zip
+
+Zipping the .tflite file is relevant in this context in order to highlight the compression benefits. Clustering does not reduce the size of the weights per se, but it renders them suitable for compression via Huffman coding and similar techniques. In consequence, though the clustered .tflite file will have similar size to the non-clustered one, their zipped versions will greatly differ in size &mdash; as shown by the figures above.
```

Comment:
Add a "Contribution" section like the [pruning RFC](https://github.com/tensorflow/model-optimization/pull/268/files?short_path=edfd107#diff-edfd10773bb6bb7662ce4fb0b94d12a1).